desc/objectives/_omnigenity.py

```diff
@@ -501,6 +502,232 @@ def compute(self, params, constants=None):
+        )
+        self._field_transforms = get_transforms(
+            self._field_data_keys,
+            obj=field,
+            grid=grid_field,
+        )
+        self._constants = {
+            "equil_transforms": self._eq_transforms,
+            "equil_profiles": self._profiles,
+            "field_transforms": self._field_transforms,
+            "helicity": self.helicity,
+        }
+
+        timer.stop("Precomputing transforms")
+        if verbose > 1:
+            timer.disp("Precomputing transforms")
+
+        if self._normalize:
+            # average |B| on axis
+            self._normalization = jnp.mean(field.B_lm[: field.M_well])
+
+        super().build(use_jit=use_jit, verbose=verbose)
+
+    def compute(self, equil_params, field_params, constants=None):
+        """Compute omnigenity errors.
+
+        Parameters
+        ----------
+        equil_params : dict
+            Dictionary of equilibrium degrees of freedom, eg Equilibrium.params_dict
+        field_params : dict
+            Dictionary of field degrees of freedom, eg OmnigenousField.params_dict
+        constants : dict
+            Dictionary of constant data, eg transforms, profiles etc. Defaults to
+            self.constants
+
+        Returns
+        -------
+        f : ndarray
+            Omnigenity error at each node (T).
+
+        """
+        if constants is None:
+            constants = self.constants
+        eq_data = compute_fun(
+            "desc.equilibrium.equilibrium.Equilibrium",
+            self._eq_data_keys,
+            params=equil_params,
+            transforms=constants["equil_transforms"],
+            profiles=constants["equil_profiles"],
+        )
+        field_data = compute_fun(
+            "desc.magnetic_fields.OmnigenousField",
+            self._field_data_keys,
+            params=field_params,
+            transforms=constants["field_transforms"],
+            profiles={},
+        )
+
+        # TODO: move to a compute function?
+        # would need to add iota as an OmnigenousField attribute
+        M, N = constants["helicity"]
+        iota = eq_data["iota"][0]  # FIXME: assumes a single flux surface
+        matrix = jnp.where(
+            M == 0,
+            jnp.array([N - M * iota, iota / N, 0, 1 / N]),
+            jnp.array([N, M * iota / (N - M * iota), M, M / (N - M * iota)]),
+        ).reshape((2, 2))
+
+        # solve for (theta_B,zeta_B) corresponding to (eta,alpha)
+        booz = matrix @ jnp.vstack((field_data["alpha"], field_data["h"]))
+        nodes = jnp.vstack((jnp.zeros_like(booz[0, :]), booz[0, :], booz[1, :])).T
+        B_eta_alpha = jnp.matmul(
+            constants["equil_transforms"]["B"].basis.evaluate(nodes), eq_data["|B|_mn"]
+        )
+        omnigenity = B_eta_alpha - field_data["|B|_omni"]
+        weights = (self.well_weight + 1) / 2 + (self.well_weight - 1) / 2 * jnp.cos(
+            field_data["eta"]
+        )
+        return omnigenity * weights
```

Comment:
do we also need quadrature weights here? or some sort of area/volume element like `sqrt(g)`?

Reply:
I don't think so, because this is in $(\eta,\alpha)$ coordinates instead of real space. 

These collocation points are not evenly spaced across the flux surface, so in theory an area element weighting could be useful. But that would add computational complexity and it doesn't seem to be necessary in practice. Having better omnigenity at the bottom of the magnetic well might actually be better than having uniform errors across the whole surface.

Reply:
hmmm ok. You may still want something to make the error independent of grid resolution